File: src/apps/motion_expression.cpp
#include <ros/ros.h>
#include <silbot3_msgs/ExpressionStart.h>
#include <silbot3_msgs/ExpressionStatus.h>
#include <std_srvs/Empty.h>
#include <string>

using namespace std;

ros::ServiceClient startMotion;
ros::ServiceClient stopMotion;
ros::Subscriber subscriber;

int status;

void callback(const silbot3_msgs::ExpressionStatus::ConstPtr& msg)
{
  ROS_INFO("ExpressionStatus : %d", msg->expression_status);
  status = msg->expression_status;
}

void callMotion(string package, string category, string id, string sentence) {
  silbot3_msgs::ExpressionStart srv;
  srv.request.package = package;
  srv.request.category = category;
  srv.request.id = id;
  srv.request.content = sentence;

  if(startMotion.call(srv)) {
    ROS_INFO("callMotion %s , %s, %s, %s",package.c_str(), category.c_str(), id.c_str(), sentence.c_str());
  } else {
    ROS_ERROR("Failed to callMotion");
  }
  ROS_INFO("finish call");
}


int main(int argc, char** argv) {
  ros::init(argc, argv, "motion_expression_cpp");
  ros::NodeHandle n;

  startMotion = n.serviceClient<silbot3_msgs::ExpressionStart>("/silbot3_expression/start");
  stopMotion = n.serviceClient<std_srvs::Empty>("/silbot3_expression/stop");

  subscriber = n.subscribe("/silbot3_expression/status", 1000, callback);
  status = -1;
  callMotion("office","guide","1","Hello World");

  while(status != 0) {
    ros::spinOnce();
  }
  ROS_INFO("finish motion expression");

  return 0;

}

File: scripts/apps/motion_expression.py
#! /usr/bin/env python
import rospy
import threading
from concurrent.futures import ThreadPoolExecutor
import time
from silbot3_msgs.srv import ExpressionStart
from silbot3_msgs.srv import ExpressionStartRequest
from silbot3_msgs.msg import ExpressionStatus

class MotionExpression :
    def __init__(self):
        self.startMotion = rospy.ServiceProxy('/silbot3_expression/start', ExpressionStart)
        self.stopMotion = rospy.ServiceProxy('/silbot3_expression/stop', ExpressionStart)
        self.subscriber = rospy.Subscriber('/silbot3_expression/status', ExpressionStatus, self.expresionCallback)
        self.expressionCondition = threading.Condition()
        self.motionFinished = True

        rospy.loginfo("wait for motion service")
        self.startMotion.wait_for_service()
        self.stopMotion.wait_for_service()
        rospy.loginfo("Found motion service")

    def callMotion(self, package, category, id, sentence):
        req = ExpressionStartRequest()
        req.expression_type = 0
        req.package = package
        req.category = category
        req.id = id
        req.content = sentence
        res = self.startMotion(req)
        rospy.loginfo("callMotion : " + str(res))

    def waitExpressionFinish(self):
        if self.motionFinished is not True :
            self.expressionCondition.wait()

    def expresionCallback(self, msg):
        rospy.loginfo(msg)

if __name__ == "__main__" :
    rospy.init_node("motion_expression_python", anonymous=False)
    expression = MotionExpression()
    expression.callMotion("office","guide","1","Hello World")
    expression.waitExpressionFinish()
